cmd/generate-catalog/generate.go

```diff
@@ -0,0 +1,382 @@
+	}
+	var input Input
+	if err = yaml.Unmarshal(inputBytes, &input); err != nil {
+		return Configuration{}, fmt.Errorf("failed to unmarshal YAML: %v", err)
+	}
+
+	oldest, err := semver.StrictNewVersion(input.OldestSupportedVersion)
+	if err != nil {
+		return Configuration{}, fmt.Errorf("invalid oldest_supported_version %q: %v", input.OldestSupportedVersion, err)
+	}
+
+	var images []BundleImage
+	for _, img := range input.Images {
+		v, err := semver.StrictNewVersion(img.Version)
+		if err != nil {
+			return Configuration{}, fmt.Errorf("invalid version %q for image %q: %v", img.Version, img.Image, err)
+		}
+		images = append(images, BundleImage{
+			Image:   img.Image,
+			Version: v,
+		})
+	}
+
+	if err = validateImageReferences(images); err != nil {
+		return Configuration{}, err
+	}
+
+	versions := getAllVersions(images)
+	if err = validateVersionsAreSorted(versions); err != nil {
+		return Configuration{}, err
+	}
+	if err = hasGapInVersions(versions); err != nil {
+		return Configuration{}, err
+	}
+	if !slices.ContainsFunc(versions, oldest.Equal) {
+		return Configuration{}, fmt.Errorf("oldest supported version %s is not present in the list of versions", oldest)
+	}
+
+	return Configuration{
+		OldestSupportedVersion: oldest,
+		Images:                 images,
+		Versions:               versions,
+	}, nil
+}
+
+// generatePackageWithIcon creates a new "olm.package" object with an operator icon.
+func generatePackageWithIcon() (Package, error) {
+	data, err := os.ReadFile(iconFile)
+	if err != nil {
+		return Package{}, fmt.Errorf("failed to read %s: %v", iconFile, err)
+	}
+	iconBase64 := base64.StdEncoding.EncodeToString(data)
+	pkg := newPackage(stableChannelName, iconBase64)
+
+	return pkg, nil
+}
+
+func generateChannels(versions []*semver.Version) []Channel {
+	channels := make([]Channel, 0)
+	var previousYStream *semver.Version
+
+	for _, v := range versions {
+		yStream := makeYStreamVersion(v)
+		if yStream.Equal(previousYStream) {
+			// If we saw the same Y-Stream version, we already created a channel for it.
+			continue
+		}
+
+		channel := newChannel(yStream)
+		channels = append(channels, channel)
+
+		previousYStream = yStream
+	}
+	return channels
+}
+
+func generateChannelEntries(versions []*semver.Version, startingVersions []*semver.Version) []ChannelEntry {
+	channelEntries := make([]ChannelEntry, 0)
+	// We know that our catalog begins with 3.62.0. We set previousEntryVersion to 3.61.0 in order to have 3.62.0's `skipRange` consistent with others.
+	previousEntryVersion := semver.New(3, 61, 0, "", "")
+	var previousYStreamVersion *semver.Version
+
+	for _, v := range versions {
+		if v.Minor() != previousEntryVersion.Minor() {
+			previousYStreamVersion = makeYStreamVersion(previousEntryVersion)
+		}
+
+		isStartingVersion := slices.ContainsFunc(startingVersions, v.Equal)
+		e := newChannelEntry(v)
+		if !isStartingVersion {
+			e.setReplaces(previousEntryVersion)
+		}
```

Comment:
I wonder if it will make code more understandable if we are to call `e.setReplaces()` here unconditionally

```suggestion
		e := newChannelEntry(v)
		e.setReplaces(previousEntryVersion)
```

but later for each first version in each channel do `e.clearReplaces()` (or `e.setReplaces(nil)`).

This communicates the idea that the very first entry in each channel is always without `replaces:` attribute. This is basic OLM rule.

This also save us from introducing the knowledge about GraphRoots/ChannelTrees/ChannelLineages in the `generateChannelEntries()` function.